hazelcast/src/main/java/com/hazelcast/map/impl/record/Records.java

```diff
@@ -50,25 +50,97 @@ public static RecordInfo buildRecordInfo(Record record) {
     }
 
     public static Object getValueOrCachedValue(Record record, SerializationService serializationService) {
-        Object value = record.getCachedValue();
-        if (value == NOT_CACHED) {
-            value = record.getValue();
-        } else if (value == null) {
-            value = record.getValue();
-            if (shouldCache(record, value)) {
-                value = serializationService.toObject(value);
-                record.setCachedValue(value);
+        Object cachedValue = record.getCachedValue();
+        if (cachedValue == NOT_CACHED) {
+            //record does not support caching at all
+            return record.getValue();
+        }
+        for (;;) {
+            if (cachedValue instanceof Thread) {
+                //the cachedValue is either locked by another thread or it contains a wrapped thread
+                cachedValue = ThreadWrapper.unwrapOrNull(cachedValue);
+                if (cachedValue != null) {
+                    //exceptional case: the cachedValue is not locked, it just contains an instance of Thread.
+                    //this can happen when user put an instance of Thread into a map
+                    //(=it should never happen, but never say never...)
+                    return cachedValue;
+                }
+                //it looks like some other thread actually locked the cachedValue
+                Thread.yield();
```

Comment:
While we're at it, it would be even better to use the complete `BackoffIdleStrategy` to have incremental backpressure. This is currently trapped on the EE side, but it would be easy to transfer.